### PR TITLE
refactor(robot_server,api): do not use ThreadManager with ProtocolEngine

### LINKED ISSUE #8023: bug: notify-server cannot send hardware events
# Overview

When subscribing to the notify-server with the "hardware_events" topic no events are parsed to the subscriber.

# Steps to reproduce

Subscribe to the notify-server with
`python3 -m notify_server.app_sub -s tcp://<ROBOT_IP>:5555 hardware_events
`
Trigger door sensor to trigger a hardware event.

# Current behavior

No event is parsed by the notify-server. [short_api.log](https://github.com/Opentrons/opentrons/files/6737306/short_api.log) shows the following:

> Jun 28 05:19:38 OT2CEP20210408B13 opentrons-api[243]: New state: 'stopped'
Jun 28 05:19:38 OT2CEP20210408B13 opentrons-api[243]: New state: 'paused'
Jun 28 05:19:38 OT2CEP20210408B13 opentrons-api[243]: Updating the window switch status: closed
Jun 28 05:19:38 OT2CEP20210408B13 opentrons-api[243]: Errored during door state event callback
                                                      Traceback (most recent call last):
                                                        File "usr/lib/python3.7/site-packages/opentrons/hardware_control/api.py", line 130, in _update_door_state
                                                        File "usr/lib/python3.7/site-packages/robot_server/hardware_wrapper.py", line 70, in _publish_hardware_event
                                                      AttributeError: 'Depends' object has no attribute 'send_nowait'

Might be caused by [this](https://github.com/tiangolo/fastapi/issues/1881) leading to the dependency of the hardware_wrapper being invoked instead of the harware_wrapper.

# Expected behavior

hardware_wrapper should be invoked to trigger a hardware_event, which should then be parsed to all topic subscribers.

### PR DESCRIPTION
**Pull request by mcous**
## Overview

What I set out to do with this PR is two things:

1. Allow the server/API to bypass `ThreadManager` if the "Enable Protocol Engine" feature flag is enabled.
2. Incorporate some of my feedback from #8213 and replace `hardware_wrapper.py` with a `robot_server.hardware` module containing initialization and dependency logic

My intention was to avoid touching the larger questions of lifetime dependencies raised in #8213, but in consolidating the initialization / cleanup logic of the hardware API, it looks to me like this code fixes #8023. It affects #8051, but I don't think it resolves it in spirit, so let's leave it open.

In touching the `get_hardware` dependency, I was forced to touch the `get_rpc_server` dependency, and it turns out those are the only two things that need explicit cleanup. I'm not sure where this leaves #8213, but I sort of think any general lifetime dependency solutions should probably wait until:

- We have more lifetime dependencies to add
- FastAPI is upgraded to Starlette v0.16 with support for lifetime context managers (https://github.com/tiangolo/fastapi/pull/3594)

## Changelog

- Do not use ThreadManager with ProtocolEngine
    - This change breaks RPC if the feature flag is set
    - This wasn't intentional, but it's a natural result of bypassing the `ThreadManager` and its `SyncAdapter`
    - I don't think it's a problem and do not intend to fix it

## Review requests

TL;DR: run protocols with the "Enable Protocol Engine" feature flag on and off, see if anything breaks. If you're not familiar with the engine stuff, smoke testing for no regressions while the `ThreadManager` is still in play is very helpful for this PR.

- Feature flag off - still using ThreadManager
    - [x] RPC API is still good [tested by @SyntaxColoring]
    - Hardware HTTP APIs are still good
        - [x] One-off requests, like the "Home" and "Lights" buttons [tested by @SyntaxColoring]
        - [x] Existing session flows, like Deck Calibration and Pipette Offset Calibration [tested by @SyntaxColoring]
- Feature flag on - ThreadManager disabled
    - RPC API will not work
    - One-off HTTP APIs may work, but if they don't it's ok
        - [The app grays them out, except for feature flag switches. –@SyntaxColoring]
    - [x] Running protocols through the engine-based `/protocols` and `/sessions` endpoints works [tested by @SyntaxColoring]

## Risk assessment

Medium due to touching hardware initialization code and nearly complete lack of test coverage

### CODE REVIEW

File: robot-server/robot_server/service/dependencies.py [hunk 1]
```diff
@@ -102,18 +48,21 @@ async def get_session_manager(
...
     opentrons_version: typing.Union[int, constants.API_VERSION_LATEST_TYPE] = Header(
         ...,
-        description=f"The requested HTTP API version which must be at "
-        f"least '{constants.MIN_API_VERSION}' or higher. To "
-        f"use the latest version specify "
-        f"'{constants.API_VERSION_LATEST}'.",
+        description=(
+            f"The requested HTTP API version must be at least "
+            f"'{constants.MIN_API_VERSION}' or higher. To use the latest "
+            f"version unconditionally, specify {constants.API_VERSION_LATEST}'"
```
Comment (mcous) [inline robot-server/robot_server/service/dependencies.py]:
```suggestion
            f"version unconditionally, specify '{constants.API_VERSION_LATEST}'"
```

File: robot-server/robot_server/app.py [hunk 2]
```diff
@@ -59,19 +54,14 @@
...
 
 @app.on_event("shutdown")
 async def on_shutdown() -> None:
     """Handle app shutdown."""
-    s = await get_rpc_server()
-    await s.on_shutdown()
-    # Remove all sessions
-    await (await get_session_manager()).remove_all()
-    # Remove all uploaded protocols
-    (await get_protocol_manager()).remove_all()
```
Comment (mcous) [inline robot-server/robot_server/app.py:74]: These weren't doing anything, nor were they necessary for the intended behavior, based on my testing. Removed them accordingly

File: robot-server/robot_server/app.py [hunk 3]
```diff
@@ -59,19 +54,14 @@
...
+    await cleanup_rpc_server(app.state)
+    await cleanup_hardware(app.state)
```
Comment (mcous) [inline robot-server/robot_server/app.py]: `cleanup_rpc_server` may still raise due to #8155. I wonder if this makes more sense with a gather and some logs:

```suggestion
    shutdown_results = await asyncio.gather(
        cleanup_rpc_server(app.state),
        cleanup_hardware(app.state),
        return_exceptions=True,
    )
    
    shutdown_errors = [r for r in shutdown_results if isinstance(r, Exception)]
    
    for e in shutdown_errors:
        log.warning(f"Error encountered during shutdown", exc_info=e) 
```

File: robot-server/robot_server/app.py (hunk 2 above)
Comment (mcous) [reply robot-server/robot_server/app.py:74]: Oof, looks like I might be wrong here. Tests are hanging on my machine in a way that reminds me of our previous experience with `ulimit`. Put the protocol manager cleanup back in place

File: api/src/opentrons/__init__.py [hunk 4]
```diff
@@ -92,67 +92,89 @@ def _find_smoothie_file() -> Tuple[Path, str]:
...
-        return ThreadManager(API.build_hardware_simulator)
-
-    # Check if smoothie emulator is to be used
+def _get_motor_control_serial_port() -> Any:
     port = os.environ.get("OT_SMOOTHIE_EMULATOR_URI")
-    if not port:
+
+    if port is None:
         smoothie_id = os.environ.get("OT_SMOOTHIE_ID", "AMA")
-        # Let this raise an exception.
```
Comment (SyntaxColoring) [inline api/src/opentrons/__init__.py:106]: Could we leave this comment in? It helped me navigate this code at one point. IIRC this raises an `IndexError` when there's no motor controller board physically attached.

(Ideally, this would probably raise a custom `NoMotorControllerPortError` or whatever, but I don't think that's for this PR.)

File: api/src/opentrons/__init__.py [hunk 5]
```diff
@@ -92,67 +92,91 @@ def _find_smoothie_file() -> Tuple[Path, str]:
...
-    hardware = await initialize_robot()
+    use_thread_manager = ff.enable_protocol_engine() is False
+
+    if use_thread_manager:
+        thread_manager = await _create_thread_manager()
+        hardware = cast(API, thread_manager)
+    else:
+        hardware = await _create_hardware_api()
+
+    async def _blink():
```
Comment (mcous) [inline api/src/opentrons/__init__.py:165]: I feel pretty suspicious that the "button blink while initializing" may not be working in this PR

File: robot-server/robot_server/app.py [hunk 6]
```diff
@@ -59,19 +56,25 @@
...
-    # Remove all uploaded protocols
-    (await get_protocol_manager()).remove_all()
+    protocol_manager = await get_protocol_manager()
+    protocol_manager.remove_all()
+
+    shutdown_results = await asyncio.gather(
+        cleanup_rpc_server(app.state),
+        cleanup_hardware(app.state),
+        return_exceptions=True,
+    )
```
Comment (SyntaxColoring) [inline robot-server/robot_server/app.py:72]: This `gather()` makes me nervous.

* The interface of the matched factory function, `get_rpc_server()`, takes a hardware parameter. As the caller, I would assume that that hardware would need to stay valid for the entire lifetime of the RPC server, including its cleanup code.
* `cleanup_rpc_server()` in fact *does not* touch the hardware, and it *does* appear safe to call it concurrently with `cleanup_hardware()`. But you have to read the implementation of `cleanup_rpc_server()` to know that. It feels like a breach of encapsulation.

I'd be more comfortable with a simple:

```python
await cleanup_rpc_server()
await cleanup_hardware()
```

Note that this would still be buggy because an exception in `cleanup_rpc_server()` will prevent `cleanup_hardware()` from running. However, I think that's okay for this PR, because:

* That interrupted shutdown bug existed before.
* That interrupted shutdown bug still exists with the `gather()` code, because of the `await get_protocol_manager()` and `protocol_manager.remove_all()` above the `gather()`.

I think solving the interrupted cleanup bug is in the domain of #8051. It probably requires either a horrible `try`/`except` mess or `contextlib.ExitStack`.

File: robot-server/robot_server/app_state.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+Mostly, this serves as a place to store singletons as well as
+in-memory databases.
+"""
+from fastapi import Request, WebSocket
+from starlette.datastructures import State as AppState
+
+
+async def get_app_state(
+    request: Request = None,
+    websocket: WebSocket = None,
```
Comment (SyntaxColoring) [inline robot-server/robot_server/app_state.py:40]: Wow. I'm amazed that this works...? Is this in the docs or discussed in a GitHub issue anywhere?

File: robot-server/robot_server/hardware.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    hardware_api = getattr(app_state, _HARDWARE_API_KEY, None)
+    unsubscribe_from_events = getattr(app_state, _HARDWARE_EVENT_UNSUBSCRIBE_KEY, None)
+
+    setattr(app_state, _HARDWARE_API_INIT_TASK_KEY, None)
+    setattr(app_state, _HARDWARE_API_KEY, None)
+    setattr(app_state, _HARDWARE_EVENT_UNSUBSCRIBE_KEY, None)
+
+    if initialize_task is not None:
+        initialize_task.cancel()
+        await asyncio.gather(initialize_task, return_exceptions=True)
```
Comment (SyntaxColoring) [inline robot-server/robot_server/hardware.py:69]: I think this can just be this?

```suggestion
        await initialize_task
```

File: robot-server/robot_server/hardware.py [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+    if initialize_task is None:
+        initialize_task = asyncio.create_task(_initialize_hardware_api(app_state))
+        setattr(app_state, _HARDWARE_API_INIT_TASK_KEY, initialize_task)
+
+
+async def cleanup_hardware(app_state: AppState) -> None:
+    """Shutdown the HardwareAPI singleton and remove it from global state."""
+    initialize_task = getattr(app_state, _HARDWARE_API_INIT_TASK_KEY, None)
+    hardware_api = getattr(app_state, _HARDWARE_API_KEY, None)
+    unsubscribe_from_events = getattr(app_state, _HARDWARE_EVENT_UNSUBSCRIBE_KEY, None)
```
Comment (SyntaxColoring) [inline robot-server/robot_server/hardware.py]: I think this is good for this PR, but one of the things I was trying to work out in #8213 was how to avoid too much indirection through keys, because it subverts typechecking.

What ended up in that PR was you'd access one `DependencySet` dataclass through a key, and then access the actual things you cared about in a type-safe way through that dataclass. So the indirection was still there, but it was consolidated and mitigated.

That strategy was messy in other ways, so I don't think it's the final solution, but I do hope we can find some other way to avoid this.

File: robot-server/robot_server/hardware.py [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+    if hardware_api is not None:
+        hardware_api.clean_up()
+
+
+async def get_hardware(app_state: AppState = Depends(get_app_state)) -> HardwareAPI:
+    """Get the HardwareAPI as a route dependency.
+
+    Arguments:
+        app_state: Global app state from `app.state`. If unspecified, FastAPI's
+            dependency injection system will attempt to provide it.
```
Comment (SyntaxColoring) [inline robot-server/robot_server/hardware.py]: Requesting changes for this.

This wording might be misleading, since FastAPI's *won't* attempt to provide `app_state` if you call `get_hardware()` without providing it. (It will only provide it when `get_hardware` is used as a dependency, like `Depends(get_hardware)`.)

File: robot-server/robot_server/service/legacy/rpc/__init__.py [hunk 11]
```diff
@@ -1 +1,22 @@
...
+
+__all__ = [
+    "CALL_RESULT_MESSAGE",
+    "CALL_ACK_MESSAGE",
+    "NOTIFICATION_MESSAGE",
+    "CONTROL_MESSAGE",
+    "CALL_NACK_MESSAGE",
+    "PONG_MESSAGE",
+    "RPCServer",
+    "get_rpc_server",
```
Comment (SyntaxColoring) [inline robot-server/robot_server/service/legacy/rpc/__init__.py:20]: Requesting changes (or at least discussion) for this.

This makes me nervous because you can now do, for example, `from service.legacy import get_rpc_server`. And then, unless you read the docstring, it looks like `get_rpc_server` is just a normal function that you can call like `get_rpc_server()`.

I'm hoping for a strong naming convention that sets our FastAPI dependency functions apart from other functions. Possibilities:

* Making sure `dependencies` is always in the import path somewhere, like what we had before, so it's `service.legacy.rpc.dependencies.get_rpc_server`. I think we discussed this before, but I don't remember your thoughts on it. I personally don't mind it.
* Making it something like `service.legacy.rpc.dep_rpc_server`.

Ditto for `robot_server.hardware.get_hardware`.

**Review by SyntaxColoring — CHANGES_REQUESTED** [review]

File: robot-server/robot_server/hardware.py (hunk 8 above)
Comment (mcous) [reply robot-server/robot_server/hardware.py:69]: This is to ensure shutdown happens cleanly (read: without raising) if:

- Initialization raised for some reason
- Shutdown is requested before the hardware initialization finishes

If the stuff that needs to be cleaned up is gone, then clean up is satisfied and a raise would be counterproductive

File: robot-server/robot_server/app_state.py (hunk 7 above)
Comment (mcous) [reply robot-server/robot_server/app_state.py:40]: I found it by reading the Starlette source, where it is present in the public API of the [`HTTPConnection` base class](https://github.com/encode/starlette/blob/84e2dc7bfe35aa82a0db958519810309e73de8f5/starlette/requests.py#L58-L62):

> ```python
> class HTTPConnection(Mapping):
>     """
>     A base class for incoming HTTP connections, that is used to provide
>     any functionality that is common to both `Request` and `WebSocket`.
>     """
>
>     # ...
>
>     @property
>     def app(self) -> typing.Any:
>         return self.scope["app"]
> ```

Probably worth a docs PR into both FastAPI and Starlette

File: robot-server/robot_server/app.py (hunk 6 above)
Comment (mcous) [reply robot-server/robot_server/app.py:72]: I hear you and agree with your concerns, but I'd rather keep the `gather` for this PR because:

- We know `protocol_manager.remove_all()`, as written, won't raise
- We know RPC cleanup _will_ raise in a common condition (connection present during shutdown)
- We know that hardware cleanup and RPC cleanup, as written, can run concurrently
- If something changes in the future that means they can't run concurrently, there will be a log statement

Removing the `gather` make me more nervous than omitting it. Let's not let perfect be the enemy of good

File: robot-server/robot_server/hardware.py (hunk 9 above)
Comment (mcous) [reply robot-server/robot_server/hardware.py]: I think this would be better if I knew how to get some sort of immutable, unique [Symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol)-like object in Python for use as the key rather than a string, where the remote but real possibility of key collisions exist, but to be honest that's the extent of my concern with this.

I don't share the concern about subverting typchecking for a few reasons:

- The keys and value storage on `AppState` are entirely private implementation details of this module
- The primary value I derive from typing in scripted languages is code documentation and maybe some refactor safety
    - I think the object keys satisfy the documentation value, and I think this module is too small for any serious refactor safety concerns
    - Some type annotations on the variable declarations probably couldn't hurt, though!

In order to strictly type `AppState`, you end up needing to:

- Create an object that has to know about the implementation details of every module in the system that may want to store arbitrary global data, breaking encapsulation
- Introduce circular (type) dependencies

Those, to me, are serious downsides that introduce brittleness and coupling into the entire application, and I don't see what value it gives us in exchange. As we work with `AppState`, all data attached to `AppState` is private to the module that creates it

File: robot-server/robot_server/service/legacy/rpc/__init__.py (hunk 11 above)
Comment (mcous) [reply robot-server/robot_server/service/legacy/rpc/__init__.py:20]: If there's one thing I've learned working with FastAPI, its that nothing is ever just a normal function you can call 🙃. Mostly, though, I guess I don't feel nervous about this? If someone comes in thinking they can call `get_rpc_server` function outside of `Depends` or without passing in `AppState` explicitly, they'll be disabused of that notion when whatever it was they were trying to do doesn't work at all.

(I'm not counting the server setup and teardown bugs that we definitely did not notice for a while because a) teardown doesn't really matter because the process is stopped, and b) the part of setup that was broken wasn't a production codepath nor was it ever subject to actual tests. In anything even remotely production-facing, I think we'd notice pretty quickly.)

My TL;DR here is that I feel it's reasonable to expect a developer in this codebase to understand how and where the dependency injection system is used. That's my 🌶️ take, so with that out of the way:

- I dislike putting `dependencies` in public module name
    - This opinion _generally_ holds for any module name that describes the type of thing it is rather than the scope of its logic in the application, including `models`, `constants`, `types`, etc.
- I could live with `get_hardware_dependency`, etc.

File: robot-server/robot_server/hardware.py (hunk 9 above)
Comment (SyntaxColoring) [reply robot-server/robot_server/hardware.py]: > In order to strictly type `AppState`, you end up needing to:
> 
> * Create an object that has to know about the implementation details of every module in the system that may want to store arbitrary global data, breaking encapsulation
> * Introduce circular (type) dependencies

Here's another approach that avoids those downsides.

```python
class TypedStateValue(typing.Generic[T]):
    def __init__(unique_key: str) -> None:
        # Alternatively, we might be able to have this autogenerate unique keys.
        self._unique_key = unique_key
    
    def get_from(app_state: AppState) -> T:
        return getattr(app_state, self._unique_key)

    def set_on(app_state: AppState, value: T) -> None:
        setattr(app_state, self._unique_key, value)
```

```python
_initialize_task = TypedStateValue[asyncio.Task]("_initialize_task")
_hardware_api = TypedStateValue[API]("_hardware_api")
_unsubscribe_from_events = TypedStateValue[typing.Callable[[], None]]("_unsubscribe_from_events")

async def cleanup_hardware(app_state: AppState) -> None:
    initialize_task = _initialize_task.get_from(app_state)
    hardware_api = _hardware_api.get_from(app_state)
    unsubscribe_from_events = _unsubscribe_from_events.get_from(app_state)

    # ...
```

I agree that the severity of not having typechecking is lessened by this module being small and these attributes being internal. I do still think we should strive to have typechecking when it doesn't take "too much" machinery, and I hope the sketch above falls under the threshold for "too much"?

Comment (mcous) [reply robot-server/robot_server/hardware.py]: That's a pretty neat little helper, will add it to `robot_server/app_state.py`

File: robot-server/robot_server/service/legacy/rpc/__init__.py (hunk 11 above)
Comment (SyntaxColoring) [reply robot-server/robot_server/service/legacy/rpc/__init__.py:20]: > If there's one thing I've learned working with FastAPI, its that nothing is ever just a normal function you can call

Magic is good because Everything Always Just Works, right? 🙃 

> Mostly, though, I guess I don't feel nervous about this?...

* It's certainly possible I'm just traumatized by the server setup/teardown bugs, but I'm still a bit nervous about the same thing happening again.
* Also, even if FastAPI dependency functions weren't special and error-prone in this way, I think I can see a separation-of-layers argument for keeping them apart?

I think let's keep this as-is for this PR, and open a naming convention discussion in `#python` or something.

File: robot-server/robot_server/hardware.py (hunk 8 above)
Comment (SyntaxColoring) [reply robot-server/robot_server/hardware.py:69]: Aha. I think I'd find a `try`/`except` clearer, but this works for me if you want to keep it as-is.

File: robot-server/robot_server/app_state.py (hunk 7 above)
Comment (SyntaxColoring) [reply robot-server/robot_server/app_state.py:40]: I'm more surprised by the `= None` thing working.

Comment (mcous) [reply robot-server/robot_server/app_state.py:40]: Oh shoot, yeah, that needs a comment. It needs to be `request: Request` and `websocket: WebSocket` in order for FastAPI to magically pick up the type annotations and inject the object. `request: Optional[Request]` will blow up DI.

From there, our mypy config is set to be a little more lenient than I'd like, where you're allowed to do `foo: SomeType = None` instead of `foo: Optional[SomeType] = None`. Will add comments to that effect because it's a gotcha

**Review by SyntaxColoring — APPROVED** [review]
Code changes since my last review look good, though there are still some open conversations.

Smoke tests (according to the plan in the description) passed. 🎉

File: api/src/opentrons/__init__.py (hunk 5 above)
Comment (mcous) [reply api/src/opentrons/__init__.py:165]: Looks like it's ok! To quote @SyntaxColoring 

> IIRC there is some redundancy in our light-blinking implementation. I think there’s a systemd service that blinks the lights until robot-server starts up, and then robot-server blinks the lights until it reaches some internal checkpoint, or something.

File: api/src/opentrons/__init__.py [hunk 12]
```diff
@@ -92,67 +92,91 @@ def _find_smoothie_file() -> Tuple[Path, str]:
...
-    log.info(f"Connecting to smoothie at port {port}")
+    log.info(f"Connecting to motor controller at port {port}")
+    return port
 
-    packed_smoothie_fw_file, packed_smoothie_fw_ver = _find_smoothie_file()
-    systemdd_notify()
-    hardware = ThreadManager(
+
+async def _create_thread_manager() -> ThreadManager:
+    """Build the hardware controller wrapped in a TheadManager.
```
Comment (sanni-t) [inline api/src/opentrons/__init__.py]:
```suggestion
    """Build the hardware controller wrapped in a ThreadManager.
```

File: api/src/opentrons/__init__.py [hunk 13]
```diff
@@ -92,67 +92,91 @@ def _find_smoothie_file() -> Tuple[Path, str]:
...
     """
     Initialize the Opentrons hardware returning a hardware instance.
     """
     robot_conf = robot_configs.load()
     logging_config.log_init(robot_conf.log_level)
 
-    log.info(f"API server version:  {__version__}")
+    log.info(f"API server version: {__version__}")
     log.info(f"Robot Name: {name()}")
+    systemdd_notify()
```
Comment (sanni-t) [inline api/src/opentrons/__init__.py:155]: This _might_ have to go after the hardware api initialization. Putting it here might result in  GPIO access conflict on a refresh robot. Have we tested this on an OT2R?

**Review by sanni-t — APPROVED** [review]
Very cool!! 
Just the comment about making sure hardware api initializes correctly on a refresh robot.

File: api/src/opentrons/__init__.py (hunk 13 above)
Comment (mcous) [reply api/src/opentrons/__init__.py:155]: I've only tested on non-refresh, I believe. I put it here because I thought I saw it come before hardware init in the existing code. Def worth double and triple checking

Comment (sanni-t) [reply api/src/opentrons/__init__.py:155]: Cool. I can go over the service files and test on the refresh electronics.

Comment (sanni-t) [reply api/src/opentrons/__init__.py:155]: Tested that refresh electronics start up correctly. PR is good to merge!
